File: 0506_relative_ranks.py
# https://leetcode-cn.com/problems/relative-ranks/

class Solution:
    def findRelativeRanks(self, score):
        """
        Inputs:
            score: list[int]
        Outputs:
            res: list[str]
        """

        n = len(score)

        # Record indices corresponding to the scores.
        mapping = {}
        for i in range(n):
            mapping[score[i]] = i

        rank = sorted(mapping.items(), key=lambda x:-x[0])

        res = [0 for i in range(n)]
        for i in range(n):
            if i == 0:
                res[rank[i][1]] = 'Gold Medal'
            elif i == 1:
                res[rank[i][1]] = 'Silver Medal'
            elif i == 2:
                res[rank[i][1]] = 'Bronze Medal'
            else:
                res[rank[i][1]] = str(i+1)
        return res

File: Cpp/0506_relative_ranks.cpp
// https://leetcode.cn/problems/relative-ranks/

class Solution {
public:
    vector<string> findRelativeRanks(vector<int>& score) {
        int n = score.size();
        vector<int> ranks(n);
        for (int i = 0; i < n; ++i) {
            ranks[i] = i;
        }
        sort(ranks.begin(), ranks.end(), [&](int i, int j) {
            return score[i] > score[j];
        });
        vector<string> res(n);
        for(int i = 0; i < n; ++i){
            if(i == 0) res[ranks[i]] = "Gold Medal";
            else if(i == 1) res[ranks[i]] = "Silver Medal";
            else if(i == 2) res[ranks[i]] = "Bronze Medal";
            else res[ranks[i]] = to_string(i+1);
        }
        return res;
    }
};
